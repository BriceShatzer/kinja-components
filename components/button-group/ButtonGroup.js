/* @flow */

import * as React from 'react';
import styled from 'styled-components';
import { darken } from 'polished';

import { ButtonWrapper } from '../buttons/Button';
import ButtonGroupItem from './ButtonGroupItem';
import ItemGroup from '../elements/item-group';


const ButtonGroupWrapper = styled.div`
	display: flex;
	white-space: nowrap;

	${ButtonWrapper} {
		border-radius: 0;

		&:not(:last-child) {
			margin-right: -1px;
		}

		&:first-child {
			border-radius: ${props => props.small ? '21px 0 0 21px' : '34px 0 0 34px'};
		}

		&:last-child {
			border-radius: ${props => props.small ? '0 21px 21px 0' : '0 34px 34px 0'};
		}

		&:hover {
			z-index: 1;

			+ ${ButtonWrapper} {
				border-left: 1px solid ${props => darken(0.2, props.theme.color.primary)};
			}
		}
	}
`;

type Props = {
	children: React.ChildrenArray<React.Element<typeof ButtonGroupItem>>,
	htmlElement?: string,
	onChange: string => void,
	small?: boolean
}

const ButtonGroup = (props: Props) => {
	const { small } = props;

	return (
		<ItemGroup
			htmlElement={ButtonGroupWrapper}
			childrenProps={{ small }}
			{...props}
		>
			{props.children}
		</ItemGroup>
	);
};


export default ButtonGroup;
