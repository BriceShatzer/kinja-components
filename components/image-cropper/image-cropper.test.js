import * as React from 'react';
import ReactImageCrop from 'react-image-crop';

import { shallow } from 'enzyme';

import ButtonGroup from '../button-group';
import ImageCropper from './image-cropper';

const IMAGE_SIZE = 2000;
const THRESHOLD_PERCENT = 50;

describe('Image Cropper', () => {
	let wrapper; // Our Image cropper component
	let plugin; // Third party component that does the actual cropping

	beforeEach(() => {
		wrapper = shallow(<ImageCropper
			src="foo.jpg"
			widthWarningThreshold={IMAGE_SIZE * (THRESHOLD_PERCENT / 100)}
			heightWarningThreshold={IMAGE_SIZE * (THRESHOLD_PERCENT / 100)}
		/>);
		plugin = wrapper.find(ReactImageCrop);
		// Simulate an image
		wrapper.setState({
			image: {
				naturalWidth: IMAGE_SIZE,
				naturalHeight: IMAGE_SIZE
			}
		});
	});

	it('renders the cropper itself', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('can select an area', () => {
		plugin.simulate('complete', {
			x: 20, y: 20,
			width: 50, height: 70
		});
		expect(wrapper).toMatchSnapshot();
	});

	it('shows a warning if the selected area is too small (both dimensions)', () => {
		plugin.simulate('complete', {
			x: 20,
			y: 20,
			width: THRESHOLD_PERCENT - 1,
			height: THRESHOLD_PERCENT - 1
		});
		expect(wrapper).toMatchSnapshot();
	});

	it('shows a warning if the selected area is too small (one dimension)', () => {
		plugin.simulate('complete', {
			x: 20,
			y: 20,
			width: THRESHOLD_PERCENT,
			height: THRESHOLD_PERCENT - 1
		});
		expect(wrapper).toMatchSnapshot();
	});

	it('can change aspect ratio', () => {
		wrapper.find('ButtonGroupItem').at(0).simulate('click');
		expect(wrapper).toMatchSnapshot();
	});

	it('should accept custom aspect ratios', () => {
		const aspectRatios = {
			'3:2': 3 / 2,
			'16:10': 16 / 10
		};
		wrapper = shallow(<ImageCropper aspectRatios={aspectRatios} src="foo.jpg" />);
		expect(wrapper).toMatchSnapshot();
		wrapper.find(ButtonGroup).simulate('change', '3:2');
		expect(wrapper).toMatchSnapshot();
	});

	it('doesn\'t let you crop below set amounts', () => {
		const MIN_AMOUNT = 200;
		wrapper.setProps({
			minWidth: MIN_AMOUNT,
			minHeight: MIN_AMOUNT
		});
		plugin = wrapper.find(ReactImageCrop);
		expect(wrapper).toMatchSnapshot();
	});
});
