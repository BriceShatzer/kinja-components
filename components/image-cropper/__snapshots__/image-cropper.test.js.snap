// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Cropper can change aspect ratio 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="16:9"
        label="16:9"
        selected={true}
        value="16:9"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "aspect": 1.7777777777777777,
        "height": 100,
        "width": 100,
        "x": 0,
        "y": 0,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <image-cropper__ImageCropperBottom>
    <image-cropper__ImageCropperInfo
      className="ImageCropper__info"
    >
      Selected area:
       
      2000
       x 
      2000
       px
    </image-cropper__ImageCropperInfo>
  </image-cropper__ImageCropperBottom>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper can select an area 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="16:9"
        label="16:9"
        selected={false}
        value="16:9"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "height": 70,
        "width": 50,
        "x": 20,
        "y": 20,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <image-cropper__ImageCropperBottom>
    <image-cropper__ImageCropperInfo
      className="ImageCropper__info"
    >
      Selected area:
       
      1000
       x 
      1400
       px
    </image-cropper__ImageCropperInfo>
  </image-cropper__ImageCropperBottom>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper doesn't let you crop below set amounts 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="16:9"
        label="16:9"
        selected={true}
        value="16:9"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "aspect": 1.7777777777777777,
        "height": 100,
        "width": 100,
        "x": 0,
        "y": 0,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={10}
    minWidth={10}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <image-cropper__ImageCropperBottom>
    <image-cropper__ImageCropperInfo
      className="ImageCropper__info"
    >
      Selected area:
       
      2000
       x 
      2000
       px
    </image-cropper__ImageCropperInfo>
  </image-cropper__ImageCropperBottom>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper renders the cropper itself 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="16:9"
        label="16:9"
        selected={true}
        value="16:9"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "aspect": 1.7777777777777777,
        "height": 100,
        "width": 100,
        "x": 0,
        "y": 0,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <image-cropper__ImageCropperBottom>
    <image-cropper__ImageCropperInfo
      className="ImageCropper__info"
    >
      Selected area:
       
      2000
       x 
      2000
       px
    </image-cropper__ImageCropperInfo>
  </image-cropper__ImageCropperBottom>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper should accept custom aspect ratios 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="3:2"
        label="3:2"
        selected={true}
        value="3:2"
      />
      <ButtonGroupItem
        key="16:10"
        label="16:10"
        selected={false}
        value="16:10"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "aspect": 1.5,
        "height": 100,
        "width": 100,
        "x": 0,
        "y": 0,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <p>
     
  </p>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper should accept custom aspect ratios 2`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="3:2"
        label="3:2"
        selected={true}
        value="3:2"
      />
      <ButtonGroupItem
        key="16:10"
        label="16:10"
        selected={false}
        value="16:10"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "aspect": 1.5,
        "height": 100,
        "width": 100,
        "x": 0,
        "y": 0,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <p>
     
  </p>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper shows a warning if the selected area is too small (both dimensions) 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="16:9"
        label="16:9"
        selected={false}
        value="16:9"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "height": 49,
        "width": 49,
        "x": 20,
        "y": 20,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <image-cropper__ImageCropperBottom>
    <image-cropper__ImageCropperInfo
      className="ImageCropper__info"
    >
      Selected area:
       
      980
       x 
      980
       px
    </image-cropper__ImageCropperInfo>
    <feedback__StyledFeedback
      arrow="topright"
      color="alert"
      text="Choose a bigger area if you want to use this image for social shares."
    />
  </image-cropper__ImageCropperBottom>
</image-cropper__ImageCropperWrapper>
`;

exports[`Image Cropper shows a warning if the selected area is too small (one dimension) 1`] = `
<image-cropper__ImageCropperWrapper>
  <image-cropper__ImageCropperRatioSelector>
    <ButtonGroup
      onChange={[Function]}
    >
      <ButtonGroupItem
        key="16:9"
        label="16:9"
        selected={false}
        value="16:9"
      />
      <ButtonGroupItem
        key="custom"
        label="Custom ratio"
        selected={false}
        value="custom"
      />
    </ButtonGroup>
  </image-cropper__ImageCropperRatioSelector>
  <p
    className="inline-help"
  >
    <span>
      Kinja recirculation modules always use a 16:9 crop.
    </span>
  </p>
  <ReactCrop
    crop={
      Object {
        "height": 49,
        "width": 50,
        "x": 20,
        "y": 20,
      }
    }
    crossorigin="anonymous"
    disabled={false}
    imageAlt=""
    keepSelection={false}
    maxHeight={100}
    maxWidth={100}
    minHeight={0}
    minWidth={0}
    onChange={[Function]}
    onComplete={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
    onImageError={[Function]}
    onImageLoaded={[Function]}
    src="foo.jpg"
  />
  <image-cropper__ImageCropperBottom>
    <image-cropper__ImageCropperInfo
      className="ImageCropper__info"
    >
      Selected area:
       
      1000
       x 
      980
       px
    </image-cropper__ImageCropperInfo>
    <feedback__StyledFeedback
      arrow="topright"
      color="alert"
      text="Choose a bigger area if you want to use this image for social shares."
    />
  </image-cropper__ImageCropperBottom>
</image-cropper__ImageCropperWrapper>
`;
